### DISCUSSION TITLE
Run program before configuring

### DISCUSSION THREAD

**Discussion by jasonbeach** [microsoft/vscode-cmake-tools#2403] [Q&A]
Is it possible to have the extension run an external program prior to configuring cmake?  The immediate use case is pulling in conan dependencies.  I currently have a build script that runs `conan install` before configuring cmake.  I can create a build task that runs the script, but that doesn't work with the cmake widgets down at the bottom of the vscode window. 

I assume I could add an custom target to cmake to run conan, but that seems tedious to do for multiple repos and I've tried hard to keep my cmake agnostic to conan.

**Comment by bobbrow** [comment]
There is a way to configure environment variables with a script prior to invoking cmake, but this script only runs once when you first select the Kit (and doesn't work with CMake Presets).  If you only need to run `conan install` once, you could potentially use that, but we don't currently have a feature that lets you run a script or "task" prior to configure or build.

**Reply by jasonbeach** [reply]
Is there any other way (or a canonical way) to get CMake Tools to play nicely with Conan?  Is adding option to run a script prior to configuring a feature that would be considered for a feature request?

I was just looking through CMake Tools settings and ran across `Cmake: Build Task` which says "Build using `tasks.json` instead of internal process" but that's the only documentation I could find on that...we do have a build task that does everything so this sounds intriguing.

**Reply by elahehrashedi** [reply]
`env` doesn't persist across tasks. So, while you can run a task before your build task, the `env` variables won't carry forward. 
The `Cmake: Build Task` setting doesn't do what you think. It does the build in the terminal. We may change the name of that setting to "Cmake: Build in Terminal".

**Reply by elahehrashedi** [reply]
if you want to run your own script, you can configure a task to run that script.
https://code.visualstudio.com/docs/editor/tasks

**Reply by jasonbeach** [reply]
> `env` doesn't persist across tasks. So, while you can run a task before your build task, the `env` variables won't carry forward. 

That' s not too big of an issue.  Running `conan install` leaves a bunch of artifacts in the build directory (at least for cmake) and as long as those are there, things are great

> if you want to run your own script, you can configure a task to run that script. https://code.visualstudio.com/docs/editor/tasks

yes we already have a build task defined to run that script. It would almost be nice if there was a setting where the selector for the default build target could show build tasks instead and then hitting the build button would execute that task. 

The problem is that there are essentially two interfaces to start building the code. The build tasks by hitting ctrl+shift+b and then the build buttons provided by cmake tools. The CMake tools buttons are super convenient, but may not execute all the needed steps for a successful build.

**Reply by bobbrow** [reply]
We have a `"cmake.buildTask": true` setting which will run the default build task for you.  Maybe something like this would work for you then?  (note: I haven't tested this and it might need some tweaking, but if you already have tasks configured, you might just need to enable the setting and ensure your build task is the default)

```
{
    // See https://go.microsoft.com/fwlink/?LinkId=733558
    // for the documentation about the tasks.json format
    "version": "2.0.0",
    "tasks": [
        {
            "label": "Build the active target",
            "type": "cmake",
            "command": "build",
            "group": {
                "kind": "build",
                "isDefault": true
            },
            "dependsOn":[ "conan install" ]
        },
        {
            "label": "conan install",
            "type": "shell",
            "command": "conan",
            "args": [
                "install"
            ]
        }
    ]
}
```

**Reply by jasonbeach** [reply]
This almost works. The corner case is when you're building for the first time. Conan has to actually run before cmake configures for the first time (so it can give it a toolchain file). So on the first time you're building, if you just select the CMake Build type to configure you'll get errors as that doesn't run the conan install task. But if you just hit the build button, it runs conan and then tries to build, but never runs cmake to configure the build. I suppose I could make another dependent task to run cmake, but that seems wrong to do that. That forces cmake to run every time you want to build, even if nothing has changed.  It does also run conan every time you build, which I think is good. That probably necessitates rerunning cmake every time, so maybe that is the way to go.

**Reply by jasonbeach** [reply]
I guess it appears there isn't a way to tell cmake tools to configure from the tasks? Looking here: https://github.com/microsoft/vscode-cmake-tools/issues/1234.  I guess I could run cmake directly, but just seems messy.
